x/stakeibc/keeper/icqcallbacks_validator_exchange_rate.go

```diff
@@ -76,18 +65,91 @@ func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query
+	validator.InternalShareToTokensRate = currentSharesToTokensRate
 	hostZone.Validators[valIndex] = &validator
 	k.SetHostZone(ctx, hostZone)
 
-	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator, "Validator Internal Exchange Rate: %v",
-		validator.InternalExchangeRate.InternalTokensToSharesRate))
+	// Check if the validator was slashed by comparing the exchange rate from the query
+	// with the preivously stored exchange rate
+	validatorWasSlashed := false
+	if previousSharesToTokensRate.IsNil() || previousSharesToTokensRate.IsZero() {
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Previous Validator Exchange Rate Unknown"))
+
+	} else {
+		validatorWasSlashed = !previousSharesToTokensRate.Equal(currentSharesToTokensRate)
+
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Previous Validator Exchange Rate: %v, Current Validator Exchange Rate: %v",
+			previousSharesToTokensRate, currentSharesToTokensRate))
+	}
+
+	// Determine if we're in a callback for the LSMLiquidStake by checking if the callback data is non-empty
+	// If this query was triggered manually, the callback data will be empty
+	inLSMLiquidStakeCallback := len(query.CallbackData) != 0
+
+	// If the validator was not slashed, and we're NOT in the callback of an LSM Liquid stake,
+	// we can stop here - there's no need to query for the delegator shares
+	if !validatorWasSlashed && !inLSMLiquidStakeCallback {
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Validator was not slashed"))
+		return nil
+	}
+
+	// If we are in an LSMLiquidStake callback, unmarshal the callback data
+	var lsmLiquidStake types.LSMLiquidStake
+	var lsmTokenDeposit recordstypes.LSMTokenDeposit
+	if inLSMLiquidStakeCallback {
+		var callbackData types.ValidatorExchangeRateQueryCallback
+		if err := proto.Unmarshal(query.CallbackData, &callbackData); err != nil {
+			return errorsmod.Wrapf(err, "unable to unmarshal validator exchange rate callback data")
+		}
+		lsmLiquidStake = *callbackData.LsmLiquidStake
+		lsmTokenDeposit = *lsmLiquidStake.Deposit
+	}
+
+	// If the validator was not slashed, and we're in the callback of an LSM liquid stake,
+	// finish the transaction to mint the user their stTokens
+	if !validatorWasSlashed && inLSMLiquidStakeCallback {
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Validator was not slashed, finishing LSM liquid stake"))
+
+		async := true
+		if err := k.FinishLSMLiquidStake(ctx, lsmLiquidStake, async); err != nil {
+			// If finishing the transaction failed, emit an event, but swallow the error so
+			// the query object still gets deleted
+			errorMessage := fmt.Sprintf("lsm liquid stake callback failed after slash query: %s", err.Error())
+			EmitFailedLSMLiquidStakeEvent(ctx, hostZone, lsmTokenDeposit, errorMessage)
+			k.Logger(ctx).Error(errorMessage)
```

Comment:
Swallowin the error makes me a bit uncomfortable, feels like it could lead to some unexpected behavior. For example, what if inside FinishLSMLiquidStake we see an error immediately after `SendCoinsFromModuleToAccount()...`. Coins would be send but LSM token wouldn't be IBC'd. Are you sure it's safe to exit from FinishLSMLiquidStake with any possible error without reverting state changes?

Reply:
Yeah I actually realized this last night as well. Moving this to slack since we already have a thread on this

Reply:
Resolved on Slack